client/events.go

```diff
@@ -0,0 +1,96 @@
+package lxd
+
+import (
+	"fmt"
+	"sync"
+)
+
+// The EventListener struct is used to interact with a LXD event stream
+type EventListener struct {
+	r            *ProtocolLXD
+	chActive     chan bool
+	disconnected bool
+	err          error
+
+	targets     []*EventTarget
+	targetsLock sync.Mutex
+}
+
+// The EventTarget struct is returned to the caller of AddHandler and used in RemoveHandler
+type EventTarget struct {
+	function func(interface{})
+	types    []string
+}
+
+// AddHandler adds a function to be called whenever an event is received
+func (e *EventListener) AddHandler(types []string, function func(interface{})) (*EventTarget, error) {
+	if function == nil {
+		return nil, fmt.Errorf("A valid function must be provided")
+	}
+
+	// Handle locking
+	e.targetsLock.Lock()
+	defer e.targetsLock.Unlock()
+
+	// Create a new target
+	target := EventTarget{
+		function: function,
+		types:    types,
+	}
+
+	// And add it to the targets
+	e.targets = append(e.targets, &target)
+
+	return &target, nil
+}
+
+// RemoveHandler removes a function to be called whenever an event is received
+func (e *EventListener) RemoveHandler(target *EventTarget) error {
+	if target == nil {
+		return fmt.Errorf("A valid event target must be provided")
+	}
+
+	// Handle locking
+	e.targetsLock.Lock()
+	defer e.targetsLock.Unlock()
+
+	// Locate and remove the function from the list
+	for i, entry := range e.targets {
+		if entry == target {
+			e.targets = append(e.targets[:i], e.targets[i+1:]...)
```

Comment:
Since the target slice contains pointers using the `append()` way of deleting leads to potential memory leaks since some elements with values might still be referenced. To avoid this go suggests using:

```
copy(e.targets[i:], e.targets[i+1:])
e.targets[len(e.targets)-1] = nil // or the zero value of T
e.targets = e.targets[:len(e.targets)-1]
```